### PR TITLE
[ONNX] Add Squeeze/Unsqueeze dynamic dimensions support when opset >= 13

### PR DESCRIPTION
**Pull request by hwangdeyu**
* Add Squeeze/Unsqueeze dynamic axes support when opset >= 13

### CODE REVIEW

**Comment by BowenBao** [comment]
Please check related test failure.

**Comment by thiagocrepaldi** [comment]
@hwangdeyu check CI

**Comment by hwangdeyu** [comment]
> @hwangdeyu check CI

Thanks for reminding, there is an issue that `squeeze` op running in ort failed, I will update this PR once I find a good solution.

**Comment by hwangdeyu** [comment]
> @hwangdeyu check CI

Thanks for reminding, there is an issue that `squeeze` op running in ort failed. I will update this PR once I find a good solution.

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -408,18 +408,35 @@ def _interpolate_warning(interpolate_mode):
...
+            return g.op("Squeeze", input, axes)
+        else:
+            return g.op("Squeeze", input, axes_i=axes_i)
+    else:  # Tensor type
+        if _export_onnx_opset_version >= 13:
+            axes_t = axes_i[0]
+            if _get_tensor_sizes(axes_t) and _get_tensor_sizes(axes_t) == 1:
+                return g.op("Squeeze", input, axes_t)
+            else:
+                raise ValueError("For squeeze axses as input, the dimension must be one in ONNX spec")
```
Comment (hwangdeyu) [inline torch/onnx/symbolic_helper.py:437]: In ONNX spec [squeeze-13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#squeeze-13), `squeeze` axses dimension must be one. 
So it doesn't easy to add the tests. Cause If we add dim as input, just like unsqueeze test in this PR. The tensor block input `prim::Param` dimension will be `0` and throw the error.

**Comment by hwangdeyu** [comment]
@BowenBao  Could you take a review of this PR?

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:437]: can we check if it is scalar, and if so apply unsqueeze to make it rank 1?

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -408,18 +408,35 @@ def _interpolate_warning(interpolate_mode):
...
+    if _is_constant(axes_i[0]):
+        if _export_onnx_opset_version >= 13:
+            axes = g.op("Constant", value_t=torch.tensor(axes_i, dtype=torch.long))
+            return g.op("Squeeze", input, axes)
+        else:
+            return g.op("Squeeze", input, axes_i=axes_i)
+    else:  # Tensor type
+        if _export_onnx_opset_version >= 13:
+            axes_t = axes_i[0]
+            if _get_tensor_sizes(axes_t) and _get_tensor_sizes(axes_t) == 1:
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:434]: `_get_tensor_sizes(axes_t) == 1` this is incorrect. probably should be `len(_get_tensor_sizes(axes_t)) == 1`

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:437]: it's not `dimension must be one``, it's `rank must be one`.

File: torch/onnx/symbolic_helper.py (hunk 2 above)
Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: change it to `_get_tensor_rank`.

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -408,18 +408,40 @@ def _interpolate_warning(interpolate_mode):
...
+            if axes_rank == 0:
+                # axes is a scalar. Unsqueeze it to a rank 1 tensor.
+                axes_t = _unsqueeze_helper(g, axes_t, [0])
+                return g.op("Squeeze", input, axes_t)
+            elif axes_rank == 1:
+                return g.op("Squeeze", input, axes_t)
+            else:
+                raise ValueError("For squeeze axses as input, the rank must be one in ONNX spec")
+        else:
+            raise ValueError("Unsupported Squeeze dynamic axes ONNX opset version: " + str(_export_onnx_opset_version))
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:444]: Make the error message actionable for the user by explaining how they can fix the problem.

```suggestion
            raise ValueError(f"opset version must be >= 13 for Squeeze with dynamic axes.")
```

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -408,18 +408,40 @@ def _interpolate_warning(interpolate_mode):
...
+            axes = g.op("Constant", value_t=torch.tensor(axes_i, dtype=torch.long))
+            return g.op("Unsqueeze", input, axes)
+        else:
+            return g.op("Unsqueeze", input, axes_i=axes_i)
+    else:  # Tensor type
+        if _export_onnx_opset_version >= 13:
+            axes_t = axes_i[0]
+            return g.op("Unsqueeze", input, axes_t)
+        else:
+            raise ValueError("Unsupported Unsqueeze dynamic axes ONNX opset version: " + str(_export_onnx_opset_version))
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:422]: Make the error message actionable for the user by explaining how they can fix the problem.

```suggestion
            raise ValueError(f"opset version must be >= 13 for Unsqueeze with dynamic axes.")
```

**Review by garymm — CHANGES_REQUESTED** [review]

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: Done.

File: torch/onnx/symbolic_helper.py (hunk 4 above)
Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: Fixed, I guess all the comments are addressed.

Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: Hi @garymm  The requested changes is fixed. Could you check it?

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -477,18 +477,40 @@ def _interpolate_warning(interpolate_mode):
...
-        return g.op("Unsqueeze", input, axes)
-    else:
-        return g.op("Unsqueeze", input, axes_i=axes_i)
+    if _is_constant(axes_i[0]):
+        if _export_onnx_opset_version >= 13:
+            axes = g.op("Constant", value_t=torch.tensor(axes_i, dtype=torch.long))
+            return g.op("Unsqueeze", input, axes)
+        else:
+            return g.op("Unsqueeze", input, axes_i=axes_i)
+    else:  # Tensor type
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:486]: decrease indentation by removing the `else` block. Since every path through the `if` block returns, no need for an explicit `else`.

File: torch/onnx/symbolic_helper.py [hunk 6]
```diff
@@ -477,18 +477,40 @@ def _interpolate_warning(interpolate_mode):
...
+        if _export_onnx_opset_version >= 13:
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:487]: Handling errors first, and then remove else block. This makes the code a bit easier to read IMO: all error handling happens up front, so we know the invariants when we get to the rest of the code. Also decreases indentation.

```python
if _export_onnx_opset_version < 13:
  raise ValueError("Opset version must be >= 13 for Unsqueeze with dynamic axes.")
return g.op("Unsqueeze", input, axes_i[0])
```

File: torch/onnx/symbolic_helper.py [hunk 7]
```diff
@@ -477,18 +477,40 @@ def _interpolate_warning(interpolate_mode):
...
+            axes_t = axes_i[0]
+            return g.op("Unsqueeze", input, axes_t)
+        else:
+            raise ValueError("Opset version must be >= 13 for Unsqueeze with dynamic axes.")
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:491]: can we provide the source range of `input` in the error message?

File: torch/onnx/symbolic_helper.py [hunk 8]
```diff
@@ -477,18 +477,40 @@ def _interpolate_warning(interpolate_mode):
...
 
 def _squeeze_helper(g, input, axes_i):
-    if _export_onnx_opset_version >= 13:
-        axes = g.op("Constant", value_t=torch.tensor(axes_i, dtype=torch.long))
-        return g.op("Squeeze", input, axes)
-    else:
-        return g.op("Squeeze", input, axes_i=axes_i)
+    if _is_constant(axes_i[0]):
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:494]: basically all the same comments apply to this function as applied to the _unsqueeze_helper above:
1. Use implicit `else` to decrease indentation
2. Handle errors first.
3. Add source range to error messages.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 9]
```diff
@@ -1525,6 +1525,16 @@ def forward(self, x):
         x = torch.randn(2, 1, 4)
         self.run_test(Squeeze(), x)
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_squeeze_dynamic_axes(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1529]: I think this should be more clear if it were named `_dynamic_dim` not `dynamic_axes`. The arg to squeeze is named `dim`, and `dynamic_axes` usually refers to the argument to `torch.onnx.export`

Please also edit the PR title and commit messages accordingly.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 10]
```diff
@@ -1533,6 +1543,16 @@ def forward(self, x):
         x = torch.randn(2, 3, 4)
         self.run_test(Unsqueeze(), x)
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_unsqueeze_dynamic_axes(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1547]: ditto on test naming

**Review by garymm — APPROVED** [review]
Feel free to merge once you've addressed all comments.

File: torch/onnx/symbolic_helper.py (hunk 8 above)
Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py:495]: Pretty helpful comments, thanks!

File: torch/onnx/symbolic_helper.py (hunk 7 above)
Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: Is it meaning we can add the input axes range? I added the message as follow:
```
f"And accepted input axes range is [{-input_rank}, {input_rank-1}].
```

File: torch/onnx/symbolic_helper.py [hunk 11]
```diff
@@ -476,19 +476,43 @@ def _interpolate_warning(interpolate_mode):
...
-    else:
+    if _is_constant(axes_i[0]):
+        if _export_onnx_opset_version >= 13:
+            axes = g.op("Constant", value_t=torch.tensor(axes_i, dtype=torch.long))
+            return g.op("Squeeze", input, axes)
         return g.op("Squeeze", input, axes_i=axes_i)
+    # Tensor type
+    input_rank = _get_tensor_rank(input)
+    if _export_onnx_opset_version < 13:
+        print("_get_tensor_rank(axes_t):", input_rank)
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:503]: remove

File: torch/onnx/symbolic_helper.py [hunk 12]
```diff
@@ -476,19 +476,43 @@ def _interpolate_warning(interpolate_mode):
...
+        raise ValueError("Opset version must be >= 13 for Squeeze with dynamic axes. " +
+                         f"And accepted input axes range is [{-input_rank}, {input_rank-1}].")
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:505]: I meant to include `input.node().sourceRange()` in the error message (if it's not empty)
On reflection we should probably add this universally for any symbolic function that fails, so don't add it here.
But for now just remove the second sentence here, go back to `"Opset version must be >= 13 for Squeeze with dynamic axes."`

Comment (hwangdeyu) [reply torch/onnx/symbolic_helper.py]: Added it. In my local test case, the input sourceRange is None, but it does not affect the error message show.
